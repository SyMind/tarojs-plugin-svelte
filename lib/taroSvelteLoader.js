const svelteLoader = require('svelte-loader')
const { componentConfig } = require('@tarojs/webpack5-runner/dist/template/component')

function transformNode(code) {
	return Promise.all([import ('code-red'), import('estree-walker')]).then(([codeRed, walker]) => {
		const program = codeRed.parse(code, {
			sourceType: 'module',
			ecmaVersion: 12,
			locations: true
		})

		let element, listen
		walker.walk(program, {
			enter(node) {
				if (node.type === 'ImportDeclaration' && node.source.type === 'Literal' && node.source.value === 'svelte/internal') {
					for (const specifier  of node.specifiers) {
						if (specifier.type === 'ImportSpecifier') {
							if (specifier.imported.name === 'element') {
								element = specifier.local.name
							}
							if (specifier.imported.name === 'listen') {
								listen = specifier.local.name
							}
						}
					}
				}

				if (node.type === 'CallExpression' && node.callee.type === 'Identifier') {
					if (node.callee.name === element) {
						const arg = node.arguments[0]
						if (arg.type === 'Literal' && arg.value.startsWith('taro-')) {
							if (process.env.TARO_ENV === 'h5') {
								arg.value = arg.value + '-core'
							} else {
								// 收集小程序模板中需要渲染的组件
								componentConfig.includes.add(arg.value)

								arg.value = arg.value.substring(5)
							}
							arg.raw = JSON.stringify(arg.value)
						}
					}
					if (node.callee.name === listen){
						const arg = node.arguments[1]
						if (arg.type === 'Literal' && arg.value === 'tap') {
							if (process.env.TARO_ENV === 'h5') {
								arg.value = 'click'
								arg.raw = JSON.stringify(arg.value)
							}
						}
					}
				}
			}
		})
		if (element || listen) {
			code = codeRed.print(program).code
		}

		return code
	})
}

module.exports = function (source) {
	const originAsync = this.async

	this.async = () => {
		const callback = originAsync.call(this)
		return (err, code, map) => {
			if (err) {
				callback(err)
			} else {
				transformNode(code).then(res => {
					callback(null, res, map)
				})
			}
		}
	}

	svelteLoader.call(this, source)
}
