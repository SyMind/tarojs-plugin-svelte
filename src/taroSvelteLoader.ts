import svelteLoader from 'svelte-loader'
import * as codeRed from 'code-red'
import * as walker from 'estree-walker'

let componentConfig
try {
	componentConfig = require('@tarojs/webpack5-runner/dist/template/component').componentConfig
} catch {
	// 兼容 Taro 3.6.5 版本
	componentConfig = require('@tarojs/webpack5-runner/dist/utils/component').componentConfig
}

if (!componentConfig) {
	throw new Error('The plugin does not support the current version of Taro')
}

function transformNode(code) {
	const program = codeRed.parse(code, {
		sourceType: 'module',
		ecmaVersion: 12,
		locations: true
	})

	let element, listen
	walker.walk(program, {
		enter(n) {
			const node = n as any

			if (node.type === 'ImportDeclaration' && node.source.type === 'Literal' && node.source.value === 'svelte/internal') {
				for (const specifier  of node.specifiers) {
					if (specifier.type === 'ImportSpecifier') {
						if (specifier.imported.name === 'element') {
							element = specifier.local.name
						}
						if (specifier.imported.name === 'listen') {
							listen = specifier.local.name
						}
					}
				}
			}

			if (node.type === 'CallExpression' && node.callee.type === 'Identifier') {
				if (node.callee.name === element) {
					const arg = node.arguments[0]
					const prefix = 't-'
					if (arg.type === 'Literal' && arg.value.startsWith(prefix)) {
						if (process.env.TARO_ENV === 'h5') {
							const tag = arg.value.substring(prefix.length)
							arg.value = 'taro-' + tag + '-core'
						} else {
							arg.value = arg.value.substring(prefix.length)

							// 收集小程序模板中需要渲染的组件
							componentConfig.includes.add(arg.value)
						}
						arg.raw = JSON.stringify(arg.value)
					}
				}
				if (node.callee.name === listen){
					const arg = node.arguments[1]
					if (arg.type === 'Literal' && arg.value === 'tap') {
						if (process.env.TARO_ENV === 'h5') {
							arg.value = 'click'
							arg.raw = JSON.stringify(arg.value)
						}
					}
				}
			}
		}
	})
	if (element || listen) {
		code = codeRed.print(program).code
	}

	return code
}

function taroSvelteLoader(source) {
	const originAsync = this.async

	this.async = () => {
		const callback = originAsync.call(this)
		return (err, code, map) => {
			if (err) {
				callback(err)
			} else {
				try {
					const res = transformNode(code)
					callback(null, res, map)
				} catch (err) {
					callback(err)
				}
			}
		}
	}

	svelteLoader.call(this, source)
}

module.exports = taroSvelteLoader
