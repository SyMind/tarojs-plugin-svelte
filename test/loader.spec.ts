import taroSvelteLoader from '../lib/taroSvelteLoader'

jest.mock('@tarojs/webpack5-runner/dist/utils/component', () => ({
    componentConfig: {
        includes: new Set()
    }
}))

function testLoader(source: string, callback: (err: Error | null, code: string, map: string) => void) {
    return done => {
        const addedDependencies = new Set();

        function cb(...args) {
            while (args.length < 4) {
                args.push(undefined);
            }
            args.push(addedDependencies);
            try {
                // @ts-ignore
                callback(...args);
            } catch (err) {
                expect(callbackSpy).toBeCalled();
                return done(err);
            }
            expect(callbackSpy).toBeCalled();
            done();
        }

        const cacheableSpy = jest.fn();

        const callbackSpy = jest.fn(cb);

        const dependencySpy = jest.fn(function(p) { addedDependencies.add(p); });

        taroSvelteLoader.call(
            {
                cacheable: cacheableSpy,
                async: () => callbackSpy,
                addDependency: dependencySpy,
                resourcePath: '<nil>'
            },
            source,
            null
        );

        expect(cacheableSpy).toBeCalled();

        for (const call of dependencySpy.mock.calls) {
            expect(typeof call[0]).toBe('string')
        }
    };
}

describe('mini program', () => {
    beforeAll(() => {
        process.env.TARO_ENV = 'weapp'
    })

    afterAll(() => {
        process.env.TARO_ENV = undefined
    })

    it('should compile', testLoader(
        '<t-view>hello, world</t-view>',
        function(err, code, map) {
            expect(err).toBeFalsy();
            expect(code).toBeTruthy();
            expect(map).toBeTruthy();
            expect(code).toMatchSnapshot();
        }
    ))
})

describe('h5', () => {
    beforeAll(() => {
        process.env.TARO_ENV = 'h5'
    })

    afterAll(() => {
        process.env.TARO_ENV = undefined
    })

    it('should compile', testLoader(
        '<t-view>hello, world</t-view>',
        function(err, code, map) {
            expect(err).toBeFalsy();
            expect(code).toBeTruthy();
            expect(map).toBeTruthy();
            expect(code).toMatchSnapshot();
        }
    ))
})
