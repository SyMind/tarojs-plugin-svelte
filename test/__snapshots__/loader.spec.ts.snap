// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`h5 should compile 1`] = `
"/* <nil> generated by Svelte v3.58.0 */
import {
	SvelteComponent,
	detach,
	element,
	init,
	insert,
	noop,
	safe_not_equal
} from "svelte/internal";

function create_fragment(ctx) {
	let t_view;

	return {
		c() {
			t_view = element("taro-view-core");
			t_view.textContent = "hello, world";
		},
		m(target, anchor) {
			insert(target, t_view, anchor);
		},
		p: noop,
		i: noop,
		o: noop,
		d(detaching) {
			if (detaching) detach(t_view);
		}
	};
}

class U3Cnilu3E extends SvelteComponent {
	constructor(options) {
		super();
		init(this, options, null, create_fragment, safe_not_equal, {});
	}
}

export default U3Cnilu3E;"
`;

exports[`mini program should compile 1`] = `
"/* <nil> generated by Svelte v3.58.0 */
import {
	SvelteComponent,
	detach,
	element,
	init,
	insert,
	noop,
	safe_not_equal
} from "svelte/internal";

function create_fragment(ctx) {
	let t_view;

	return {
		c() {
			t_view = element("view");
			t_view.textContent = "hello, world";
		},
		m(target, anchor) {
			insert(target, t_view, anchor);
		},
		p: noop,
		i: noop,
		o: noop,
		d(detaching) {
			if (detaching) detach(t_view);
		}
	};
}

class U3Cnilu3E extends SvelteComponent {
	constructor(options) {
		super();
		init(this, options, null, create_fragment, safe_not_equal, {});
	}
}

export default U3Cnilu3E;"
`;
